dali/image/tiff.cc

```diff
@@ -80,15 +81,20 @@ Image::Shape TiffImage::PeekShapeImpl(const uint8_t *encoded_buffer, size_t leng
       } else if (tag_id == HEIGHT_TAG) {
         height = value;
         height_read = true;
-      } else if (tag_id == SAMPLESPERPIXEL_TAG) {
+      } else if (tag_id == SAMPLESPERPIXEL_TAG && !photometric_read) {
         nchannels = value;
         nchannels_read = true;
+      } else if (tag_id == PHOTOMETRIC_INTERPRETATION_TAG && value == PHOTOMETRIC_PALETTE) {
+        nchannels = 3;
+        photometric_read = true;
       }
     }
+    if (width_read && height_read && nchannels_read && photometric_read)
```

Comment:
If we read the PHOTOMETRIC_INTERPRETATION_TAG, we automatically set the nchannels = 3. Do we still need to search for the nchannels value? The enforce below assumes alternative between reading nchannels and the photometric tag.

Reply:
That's right, once we have `photometric`, we don't have to look for `nchannels`. But even if we have `nchannels`, we have to look for `photometric`, which doesn't have to be present at all. So the only way to early-exit the loop should be:
```cpp
if (width_read && height_read && photometric_read)
```
In other words, enforcing `nchannels_read` makes no sense, because even if we do have `nchannels`, we still have to look for possible `photometric`.

I'll change the code in a moment.

Reply:
Actually you need to check that one of the two is true:
1. photometric read and it's palette
2. samples per pixel tag is read

How about using `nchannels_read`?
```
if (tag_id == SAMPLESPERPIXEL_TAG) {
  assert(!nchannels_read);
  nchannels = value;
  nchannels_read = true;
} else if (tag_id == PHOTOMETRIC_INTERPRETATION_TAG && value == PHOTOMETRIC_PALETTE) {
  assert(!nchannels_read);
  nchannels = 3;
  nchannels_read = true;
}
...
if (width_read && height_read && nchannels_read) {
  ...
}
```

Reply:
@jantonguirao, most palette tiffs have tags like this (in any order):
```
SAMPLESPERPIXEL_TAG = 1
PHOTOMETRIC_INTERPRETATION_TAG = PHOTOMETRIC_PALETTE
```
They have `SAMPLESPERPIXEL_TAG = 1` because, technically speaking, there is 1 number for each pixel. So even if we get samples per pixel, there still might be `PHOTOMETRIC_INTERPRETATION_TAG` that would impact the actual number of channels. So, breaking immediately after reading samples per pixel would be wrong

Reply:
@klecki suggested to add a comment explaining it, I'll do it in a moment

Reply:
In https://github.com/NVIDIA/DALI/pull/4089/commits/0052563285dca7c98bbb93aa5a4cfc66b9a00723 I've added a short comment explaining this, and also renamed the bools to clarify stuff:
- **Renamed `nchannels_read` to `samples_per_px_read`**, because that one was set when `SAMPLESPERPIXEL_TAG` was encountered. Also, it was possible that `nchannels` was set by the `PHOTOMETRICINTERPRETATION_TAG` and `nchannels_read` was still `false`. Quite confusing...
- **Renamed `photometric_read` to `palette_read`**, because it was only set if `PHOTOMETRICINTERPRETATION_TAG` was equal to `PALETTE`, and not for any `PHOTOMETRICINTERPRETATION_TAG`